### ISSUE TITLE
feat(http): accept string for method in RequestOptions class and interface

### ISSUE THREAD

**Issue by jeffbcross** [angular/angular#2840]
Currently, the method must be of the `ResponseMethods` TypeScript enum, which is really an integer. So to create a request in TypeScript, one would use the enum to describe the method instead of passing in the method string:

```typescript
var options = new RequestOptions({
  url: '/foo.json',
  method: RequestMethods.POST
});
var request = new Request(options);
```

This adds some convenience for TypeScript and Dart users to get some IDE/compiler help to make sure the method is a valid value. But plain old ES5 users probably don't want to pass `0` as the method value. Instead, `RequestOptions` should have the `method` property be a union type of `RequestMethods | string`, and should validate the provided string & throw if not a recognized method.

I'm not sure `Request` class needs to accept method as a string though. Ideally, we could reduce instances of type ambiguity to public APIs.

**Comment by mgechev** [comment]
Since `Request` is exposed by the public API, it may create confusion if the users can pass both strings and `RequestMethods` to the `http` methods when using object literals and only `RequestMethods` when using `Request`.

@jeffbcross I added support of passing HTTP method names as strings to `RequestOptions` [here](https://github.com/mgechev/angular/commit/d363256de1876f8abaf381de945957084fbad6ec). Do you think it looks okay and ready for PR?

**Comment by jeffbcross** [comment]
@mgechev I left a couple of comments on the commit, but yeah, looks good! Please PR! I will assign this to myself just to show that someone's working on it (github issues can only be assigned to those with commit access).

**Comment by jeffbcross** [comment]
> Since Request is exposed by the public API, it may create confusion if the users can pass both strings and RequestMethods to the http methods when using object literals and only RequestMethods when using Request.

I agree, please also add the `normalizeMethod` to `Request`. `Request` actually is a useful public API, especially in tests.

**Comment by mgechev** [comment]
I took another look at `Request`. It accepts `RequestOptions`, however, inside the `RequestOptions`'s constructor we already normalize the method name. Which means the following is valid:

```
http.request(new Request(new RequestOptions({url: 'https://google.com', method: 'post' })));
```

However, in case the `Request` constructor function is used with ES5 it won't work. I added one extra call of `normalizeMethodName` inside `Request`'s constructor.

Here is the [PR](https://github.com/angular/angular/pull/4331).

**Comment by jeffbcross** [comment]
@mgechev eventually `Request` should accept `RequestOptionsArgs` interface, but that's not in scope for this PR. I think I removed that when I was making `Http` work for dart some time ago.

**Comment by mgechev** [comment]
@jeffbcross I updated the PR. If declaring that the argument of the constructor of `Request` should be of type `RequestOptionsArgs` is out of the scope of this PR I can open another one?

**Comment by jeffbcross** [comment]
Closed via 34518f0f2d23019f34829bb49eb3d3f4efe80516
